### PR TITLE
[TA2299] Fix replica registration bug

### PR DESCRIPTION
**Pull request by utkarshmani1997**
On branch fix-replica-registration-bug
Changes to be committed:
	modified:   controller/control.go
	modified:   controller/rest/model.go
	modified:   replica/rest/replica.go

1. Why is this change necessary?
-  Controller was trying to connect with the first registered replica only, when
   that replica gets crashed before getting start signal.
-  Controller was not caching the error in case when 'start' signal was not sent
   successfully.
-  Need for better and descriptive logs while registration of replica.

2. How does it address the issue?
-  This commit catch the error after signalling replicas and sets
   c.MaxRevReplica = "" and also deletes the entry for this replica from map.
-  Adds descriptive logs.
-  Isolated the debug code into separate file

3. What side effects does this change have?
-  None

Signed-off-by: Utkarsh Mani Tripathi <utkarshmani1997@gmail.com>

### CODE REVIEW

File: controller/control.go [hunk 1]
```diff
@@ -266,9 +266,15 @@ func (c *Controller) registerReplica(register types.RegReplica) error {
 		if c.MaxRevReplica == register.Address {
 			logrus.Infof("Replica %v signalled to start again %d:%d", c.MaxRevReplica,
 				len(c.RegisteredReplicas), c.replicaCount)
-			c.signalToAdd()
+			if err := c.factory.SignalToAdd(c.MaxRevReplica, "start"); err != nil {
+				logrus.Errorf("Replica %v is not able to send 'start' signal, found err: %s",
+					c.MaxRevReplica, err.Error())
+				delete(c.RegisteredReplicas, register.Address)
+				c.MaxRevReplica = ""
```
Comment (vishnuitta) [inline controller/control.go]: c.StartSignalled need to be set to false

File: controller/control.go [hunk 2]
```diff
@@ -288,13 +294,19 @@ func (c *Controller) registerReplica(register types.RegReplica) error {
...
+			logrus.Errorf("Replica %v is not able to send 'start' signal, found err: %s", c.MaxRevReplica, err.Error())
+			delete(c.RegisteredReplicas, register.Address)
+			c.MaxRevReplica = ""
+			return err
+		}
+		c.StartSignalled = true
 		return nil
 	}
 
+	logrus.Warning("No of yet to be registered replicas are less than ", c.replicaCount)
```
Comment (vishnuitta) [inline controller/control.go]: log the count of registered replicas also

File: backend/remote/remote.go [hunk 3]
```diff
@@ -280,6 +281,7 @@ func (rf *Factory) SignalToAdd(address string, action string) error {
 			Timeout: timeout,
 		},
 	}
+	time.Sleep(inject.SignalToAddTimeout)
```
Comment (AmitKumarDas) [inline backend/remote/remote.go]: This is ok. However, do think about other ways to inject.

How about creating a new Factory and making use of this new Factory. The original Factory remains unchanged. The enhancements & or fixes go into original Factory only. Injection related capabilities are isolated in this new Factory called `injectibleFactory`.

```go
func InjectibleFactory() types.BackendFactory {
  return &InjectingFactory{}
}

type InjectingFactory struct {
  Factory
}

// verify if InjectingFactory implements all the methods exposed by 
// types.BackendFactory
var _ types.BackendFactory = &InjectingFactory{}

func (if *InjectingFactory) SignalToAdd(address string, action string) error {
  time.Sleep(inject.SignalToAddTimeout)
  if.Factory.SignalToAdd(address, action)
}
```

Comment (AmitKumarDas) [reply backend/remote/remote.go]: In addition, the image tag can be `0.6.0-debug` & so on.

Comment (utkarshmani1997) [reply backend/remote/remote.go]: Thanks @AmitKumarDas  for the review, will change the image tag.

**Pull request merged**